pkg/runner/tunnel_client.go

```
@@ -0,0 +1,257 @@
+type streamMetadata struct {
+	RemotePort string `json:"remote_port"`
+}
+
+// maxMetadataSize is the maximum size of stream metadata to prevent DoS attacks.
+const maxMetadataSize = 1024
+
+// handleStream processes a single smux stream by spawning a worker subprocess
+// with user credentials to connect to the local service.
+func (tc *TunnelClient) handleStream(stream *smux.Stream) {
+	defer stream.Close()
+
+	// Read metadata with size limit to prevent memory exhaustion from malicious servers
+	limitedReader := &io.LimitedReader{R: stream, N: maxMetadataSize}
+	bufReader := bufio.NewReader(limitedReader)
+	metadataLine, err := bufReader.ReadString('\n')
+	if err != nil {
+		log.Debug().Err(err).Msg("Failed to read stream metadata (may exceed size limit).")
+		return
+	}
+
+	var metadata streamMetadata
+	if err := json.Unmarshal([]byte(metadataLine), &metadata); err != nil {
+		log.Debug().Err(err).Msg("Failed to parse stream metadata.")
+		return
+	}
+
+	// Use target port from tunnel configuration if not specified in metadata
+	targetPort := metadata.RemotePort
+	if targetPort == "" {
+		targetPort = fmt.Sprintf("%d", tc.targetPort)
+	}
+
+	targetAddr := fmt.Sprintf("127.0.0.1:%s", targetPort)
+
+	// Spawn tunnel worker subprocess (runs as nobody user for security)
+	cmd, stdinPipe, stdoutPipe, err := spawnTunnelWorker(targetAddr)
+	if err != nil {
+		log.Debug().Err(err).Msgf("Failed to spawn tunnel worker for %s.", targetAddr)
+		return
+	}
+
+	defer func() {
+		stdinPipe.Close()
+		if cmd.Process != nil {
+			if err := cmd.Process.Kill(); err != nil {
+				log.Debug().Err(err).Msg("Failed to kill tunnel worker process.")
+			}
+		}
+		_ = cmd.Wait()
+	}()
+
+	log.Debug().Msgf("Tunnel worker spawned for %s.", targetAddr)
+
+	// Create combined reader: remaining buffered data + original stream
+	// This is needed because bufReader may have buffered data beyond metadata
+	remainingBuf := make([]byte, bufReader.Buffered())
+	_, _ = bufReader.Read(remainingBuf)
```

Is there a specific reason why the error from `bufReader.Read()` is being ignored? 
I'm asking because ignoring this error might lead to potential loss of buffered data.

---

Great catch. I've updated the logic to use io.ReadFull to ensure all buffered data is fully captured without any loss. I also added proper error handling to maintain data integrity as suggested.